bip-wallet-policies.mediawiki

```diff
@@ -0,0 +1,308 @@
+information;
+* replacing every <tt>/**</tt> with <tt>/<0;1>/*</tt>.
+
+For example, the wallet descriptor <tt>pkh(@0/**)</tt> with key information
+<tt>["[d34db33f/44'/0'/0']xpub6ERApfZwUNrhLCkDtcHTcxd75RbzS1ed54G1LkBUHQVHQKqhMkhgbmJbZRkrgZw4koxb5JaHWkY4ALHY2grBGRjaDMzQLcgJvLJuZZvRcEL"]</tt>
+produces the following multipath descriptor:
+
+<tt>pkh([d34db33f/44'/0'/0']xpub6ERApfZwUNrhLCkDtcHTcxd75RbzS1ed54G1LkBUHQVHQKqhMkhgbmJbZRkrgZw4koxb5JaHWkY4ALHY2grBGRjaDMzQLcgJvLJuZZvRcEL/<0;1>/*)</tt>
+
+=== Implementation guidelines ===
+
+Implementations must not necessarily implement all the possible wallet policies defined by this standard, but it is recommended to clearly document any limitation.
+
+Implementations can add additional metadata that is stored together with the wallet policy for the purpose of wallet policy registration and later usage. Metadata can be vendor-specific and is out of the scope of this document.
+
+Any implementation in a software wallet that allows wallet policies not matching any of the specifications in [[bip-0044.mediawiki|BIP-44]], [[bip-0049.mediawiki|BIP-49]], [[bip-0084.mediawiki|BIP-84]], [[bip-0086.mediawiki|BIP-86]] (especially if involving external cosigners) should put great care into a process for backing up the wallet policy that represents the account. In fact, unlike standard single-signature scenarios, the seed alone is no longer enough to discover wallet policies with existing funds, and the loss of the backup is likely to lead to permanent loss of funds. Unlike the seed, leaking such backups only affects the privacy of the user, but it does not allow the attacker to steal funds.
+
+Avoiding key reuse among different wallet accounts is also extremely important, but out of scope for this document.
+
+== Examples ==
+
+In the examples in this section, the vector of key information items is omitted. See the test vectors below for complete examples.
+
+Common single-signature account patterns:
+* <tt>pkh(@0/**)</tt> (legacy).
+* <tt>wpkh(@0/**)</tt> (native segwit).
+* <tt>sh(wpkh(@0/**))</tt> (nested segwit).
+* <tt>tr(@0/**)</tt> (taproot single-signature account).
+
+Common multisignature schemes:
+* <tt>wsh(multi(2,@0,@1))</tt> - SegWit 2-of-2 multisignature, keys in order.
+* <tt>sh(sortedmulti(2,@0,@1,@2))</tt> - Legacy 2-of-3 multisignature, sorted keys.
+
+Some miniscript policies in <tt>wsh</tt>:
+* <tt>wsh(and_v(v:pk(@0),or_d(pk(@1),older(12960))))</tt> - Trust-minimized second factor, degrading to a single signature after about 90 days.
+* <tt>wsh(thresh(3,pk(@0/**),s:pk(@1/**),s:pk(@2/**),sln:older(12960)))</tt> - A 3-of-3 wallet that becomes a 2-of-3 if coins are not spent for about 90 days.
+* <tt>wsh(or_d(pk(@0),and_v(v:multi(2,@1,@2,@3),older(65535))))</tt> - A singlesig wallet with automatic inheritance to a timelocked 2-of-3 multisig of family members.
+
+== Test Vectors ==
+
+[[bip-0044.mediawiki|BIP-44]], first account
+ Descriptor template: pkh(@0)
```

Comment:
We have to also allow `/*` following `@n` (i.e. a single asterisk instead of two), since not all wallets use 0/1 paths to differentiate change addresses.

I'm OK with `@n` alone expanding to `@0/**` but this does add even more malleability to expressions. Regardless, the `always` section should be update to allow single asterisks.

Reply:
The naked `@N` was an oversight, I don't think it should be allowed as it would be confusing (Ledger's implementation doesn't permit it). I will fix it.

--

If we generalize the suffix to expressions other than `<0;1>/*`, then we have to make sure that no pubkey is repeated.
In practice, for `/*`, that would mean that _all_ the placeholders must be followed by the exact suffix `/*`.

Out of curiosity, what software wallets use that scheme?

Reply:
> If we generalize the suffix to expressions other than <0;1>/*

I'm not sure I follow. I'm not suggesting generalizing the suffix, at least in the sense of allowing general path expressions to be used as suffixes. I'm saying the only key expression suffixes that should be allowed are `/**`, `/<m;n>/*` and `/*`.

> If we generalize the suffix to expressions other than <0;1>/*, then we have to make sure that no pubkey is repeated.

I assume you mean here that a key expression could be given as `[/a/b/c]foo/**` and then elsewhere as `[/a/b/c/0]bar/*` and/or `[/a/b/c/1]bar/*` giving rise to the same derived keys with different key identifiers.

> In practice, for /*, that would mean that all the placeholders must be followed by the exact suffix /*

For a given key expression, yes. I'm fine with supporting `/*` being optional, and if present the implementation must ensure this.

Note also that at present this BIP does not state that implementations must ensure that `KP` expressions refer to distinct keys when substituting them into a policy, although that appears to be a requirement. I think it would also be clearer to state that all key expressions in a policy must be replaced by `KP` expressions (i.e. that no non-`KP` keys are allowed).

Also missing are that `<m;n>` expressions should be numerically sorted, and that explicit `<0;1>` should be disallowed.

I'm happy to make individual commits with suggested changes for the above for you to review/cherry-pick. Should I create a PR against your repo for this?.

> Out of curiosity, what software wallets use that scheme?

Blockstream Green multisig wallets do not use bip44 style paths (the wallet predates the BIP). Also descriptors themselves do not enforce bip44 derivation.

Reply:
> I'm not sure I follow. I'm not suggesting generalizing the suffix, at least in the sense of allowing general path expressions to be used as suffixes. I'm saying the only key expression suffixes that should be allowed are `/**`, `/<m;n>/*` and `/*`.

What I mean is that if we allow mixing things like  `@0/*` with `@1/<m;n>/*`, then we for the xpub of `@1` we would need to derive `@1/m` and `@1/n` and compare with the pubkey of the xpub in `@0`, which is quite expensive on a hardware wallet.

With no mixing, for any two placeholders like `@0/<m;n>/*` and `@0/<p;q>/*`, I can just check that the sets `{m, n}` and `{p, q}` are disjoint.

> For a given key expression, yes. I'm fine with supporting `/*` being optional, and if present the implementation must ensure this.

Not just _for a given key expression_, I mean it across the entire wallet policy. That is, either the KPs all end with `/*`, or they all end with some `/<NUM;NUM>/*`.

> Note also that at present this BIP does not state that implementations must ensure that `KP` expressions refer to distinct keys when substituting them into a policy, although that appears to be a requirement.

Good point, worth adding explicitly the "Additional rules", with the motivation that it's required in miniscript (and it's simpler to add this restriction globally, rather than on the individual miniscript parts of the policy).

So basically:
- all the public keys (deserialized from the corresponding xpubs in the keys information vector) must be distinct
- all the derivations in each KP expression referring to the same index in the keys information vector must be distinct

In order to allow optional extensions like `/*`, perhaps I could mention that possibility in a separate section?

> I think it would also be clearer to state that all key expressions in a policy must be replaced by `KP` expressions (i.e. that no non-`KP` keys are allowed).

A wallet policy is the pair `(descriptor_template, keys_information_vector)`, so not sure what you mean here. The grammar seems to specify exactly what are the allowed KP and KEY expressions, as far as I can tell.

> Also missing are that `<m;n>` expressions should be numerically sorted, and that explicit `<0;1>` should be disallowed.

I like suggesting to sort `<m;n>` numerically, but not sure about disallowing `<0;1>`. I expect people to prefer not mixing `@0/**` with `@0/<2;3>/*`, as it's more explicit in that case to just write `@0/<0;1>/*`.

> I'm happy to make individual commits with suggested changes for the above for you to review/cherry-pick. Should I create a PR against your repo for this?.

Sure, feel free to propose changes, but I'll try to incorporate the comments now. I'm also hoping to find the time for a general pass over the whole document to improve the Motivation section, as that was written before the BtcPrague meeting and it can surely be improved in hindsight.

Reply:
> I like suggesting to sort <m;n> numerically, but not sure about disallowing <0;1>. I expect people to prefer not mixing @0/** with @0/<2;3>/*, as it's more explicit in that case to just write @0/<0;1>/*.

What's the reason for sorting? It's not mentioned in [BIP-389](https://github.com/bitcoin/bips/blob/master/bip-0389.mediawiki).

+1 to keep allowing `<0;1>/*` for readability for the reason you stated.

Reply:
> Not just for a given key expression, I mean it across the entire wallet policy.

OK, I understand where you are coming from now. Enforcing this across the entire policy is sensible (no-one should be mixing key expressions with different cardinalities), and its pretty easy to implement.

> So basically:

Yes these 2 two points are whats missing IMO.

> perhaps I could mention that possibility in a separate section?

Sure, that works, let me know if you'd like me to draft it or if you are happy to.

> The grammar seems to specify exactly what are the allowed KP and KEY expressions, as far as I can tell.

This is true, I just prefer being more explicit. Feel free to ignore this if you disagree.

> I expect people to prefer not mixing
> +1 to keep allowing <0;1>/* for readability

No doubt its slightly better for human readability in the rare case where alternate multipath derivations are given. But its bad for comparing two such policies (or checking against a whitelist of supported policies) because it's another source of malleability. It also complicates logic to convert a descriptor to a human friendly policy automatically, because substitution of `/<0;1>/*` for `/**` as recommended in this BIP will not result in the human friendly format in this case. Complicated policies are far more likely to be automatically generated than human written, and so in practice I suspect the large majority will end up mixing `/**` and `/<M;N>/*` anyway. However its your decision.

> What's the reason for sorting?
Malleability and ease of comparing multipaths. Unfortunately 389 states that the meaning of all n=2 multipaths is fixed (receive addresses and change), and so sorting is probably not possible to enforce given some hypothetical wallet out there could use descending numbers for this purpose.

>  It's not mentioned in BIP-389
True, but we limit multipaths to length 2 in policies which is also not part of 389.

Reply:
> Unfortunately 389 states that the meaning of all n=2 multipaths is fixed (receive addresses and change), and so sorting is probably not possible 

I was actually thinking primarily about this case and wondered why one would restrict people from using receive and change indices to be ordered.

> Malleability and ease of comparing multipaths.

`/<1;2;3>*` is not the same as `<3;2;1>/*`, so why should they be compared or treated to be the same? Applications may assign meaning to the position (similar to receive and change with two elements).

In practice it probably does not matter much, but I wouldn't impose this restriction here.

> Blockstream Green multisig wallets do not use bip44 style paths (the wallet predates the BIP). 

Out of curiosity, how does Green identify which UTXOs are change?

Reply:
As per this discussion:
- 3ade41fc6bc12461388f34d5aebe9ba60a008704 makes the rules on distinct pubkeys explicit;
- 0d6e07781e17a4ee8b208b3aea5985fed7ac2cbc adds a section for the optional additional derivations, pointing out the potential footguns I could think of.

I didn't add any requirement/suggestion to sort `<m;n>` expressions. Thanks for the comments, and feel free to suggest further changes, of course!

Reply:
@bigspider Thanks!

Comments on the current state:
1. under `** a string of the form` there should be a new bullet point stating that any path expression as listed in the `Optional derivation paths` is optionally allowed.
2. There still appear to be some bare `@n` references in the test vectors (e.g. multisig).
3. The `Why must public keys be distinct?` reference doesn't appear to point anywhere. Do you have a reference to any docs/discussion on pubkey reuse in miniscript this can link to?
4. Some invalid test vectors would be helpful for implementors to check against (see below).

I'm fully implementing the BIP in wally now, and may have some further comments once that work is complete.

Suggested invalid test vectors:
1. `pkh(@0)` Key with no following path
2. `pkh(@0/0/*)` Key with an explicit path present
3. `sh(multi(1,@1/**,@0/**))` Keys used out of order
5. `sh(multi(1,@0/**,@0/**))` Repeated keys w/same path expression
6. `sh(multi(1,@0/<0;1>/*,@1/<1;2>/*))` Non-disjoint multipath expressions
7. `sh(multi(1,@0/**,xpub6AHA9hZDN11k2ijHMeS5QqHx2KP9aMBRhTDqANMnwVtdyw2TDYRmF8PjpvwUFcL1Et8Hj59S3gTSMcUQ5gAqTz3Wd8EsMTmF3DChhqPQBnU/<0;1>/*))` Expression with a non `KP` key present
8. `pkh(@0/<0;1;2>/*)` Solved cardinality > 2

Reply:
@benma 

note slightly O/T to this BIP.

> why one would restrict people from using receive and change indices to be ordered.
> /<1;2;3>* is not the same as <3;2;1>/*, so why should they be compared or treated to be the same? Applications may assign meaning to the position (similar to receive and change with two elements).

They are the same assuming no other paths appear in the same expression; they both result in the same set of solutions, just in a different order. Of course this is not true with multiple keys: `<1;2;3>` and `<4;5;6>` have different solutions to `<3;2;1>` and `<4;5;6>` due to the stepwise resolution through path elements when generating the solutions. But in the single case its a cause of expression malleability.

Note that ordering of expressions is not explicitly specified anywhere other than this stepwise derivation, and no meaning can be inferred except in the n=2 case (where all n=2 cases must be considered as receive/change even if this is not the intention). Consider expressions like `combo` or new extensions like explicit key lists. In what order is `combo(<key>/<1;2;3>/*)` solved, paths iterated first or the combo variants for each path?

As per my comments on the BIP the addition of extensions that change cardinality and the non-specification of ordering for N>3 means that we have the worst of all worlds: proscribed meaning for one case that prevents general use of N=2, and no ability to state anything about all the meaning of solutions when other expressions that change cardinality are present. Enforcing ordering and describing the expansion order of expressions would not have been onerous but would have allowed implementations to process them consistently instead of needing whitelists/understanding specific patterns.

> In practice it probably does not matter much, but I wouldn't impose this restriction here.

Agreed, this ship along with other malleability concerns has unfortunately sailed. Although it would be nice to enforce `'` rather than `h` for hardening in policies at least, @bigspider?

> Out of curiosity, how does Green identify which UTXOs are change?

For singlesig wallets we use bip44 like everyone else, although you can't rely on the internal/external paths only for change detection, since its possible to create txs using either. For multisig we don't explicitly track change, the backend computes limits etc based on the net effect on the wallet, and displaying change is based on a simple check of the txs outputs (one output only to the wallet = redeposit, non-wallet output plus wallet output = payment and change etc).

Reply:
> Agreed, this ship along with other malleability concerns has unfortunately sailed. Although it would be nice to enforce `'` rather than `h` for hardening in policies at least, @bigspider?

I agree; Ledger's implementation only allows `'` (choice mostly because it looks better on a small screen).
The current specs indeed don't mention `h`.

I'll go over the other comments (and add the failure cases, thanks!) next week.

Reply:
The spec above says that the @N must always be followed by /** or /<NUM:NUM>/*, but some examples and test vectors don't have that suffix.

 followed by a non-negative decimal number, with no leading zeros (except for <tt>@0</tt>)
* ''always'' followed by either:
** the string  <tt>/**</tt>, or
...
Is @0 the same as @0/**? Would be good to fix either the specification or the examples/test vectors.

Reply:
^ this comment is spam from a bot/scam/shill user

Reply:
4b1f826217ab8f99c9484fbd6f1b1f88fa0c3bc0 should address all the remaining comments from this thread, thanks @jgriffiths!
 
> Suggested invalid test vectors:
> 
> 5. `sh(multi(1,@0/<0;1>/*,@1/<1;2>/*))` Non-disjoint multipath expressions

This is actually valid as <tt>@0</tt> and <tt>@1</tt> are independent; added an example where both are <tt>@0</tt> instead.

Reply:
For the miniscript reference, I suppose https://bitcoin.sipa.be/miniscript/ is the only one, although the mention of separate keys is quite [buried in the page](https://bitcoin.sipa.be/miniscript/#:~:text=We%20restrict%20this,using%20separate%20keys.) in the section on malleability.

Reply:
@bigspider Thanks!

The wally implementation (https://github.com/ElementsProject/libwally-core/pull/369) is now complete and is being tested for addition to Jade. Once merged it would be good to add to the implementations section. At this stage it looks like this BIP has support in 3 HWW so getting a BIP number assigned should not be an issue.

It would be good IMO to squash all commits at this stage. I'll do a final review following wally code review and Jade testing. Many thanks for your responsiveness in addressing review comments.

Reply:
FYI policy support has been merged into wally now and support in Jade is now being tested.